src/librustc/mir/mir_cfg.rs

```diff
@@ -0,0 +1,78 @@
+// Copyright 2016 The Rust Project Developers. See the COPYRIGHT
+// file at the top-level directory of this distribution and at
+// http://rust-lang.org/COPYRIGHT.
+//
+// Licensed under the Apache License, Version 2.0 <LICENSE-APACHE or
+// http://www.apache.org/licenses/LICENSE-2.0> or the MIT license
+// <LICENSE-MIT or http://opensource.org/licenses/MIT>, at your
+// option. This file may not be copied, modified, or distributed
+// except according to those terms.
+
+use std::{iter, slice};
+use rustc_data_structures::indexed_vec::IndexVec;
+use rustc_data_structures::graph_algorithms::{Graph, GraphPredecessors, GraphSuccessors};
+use mir::repr::{Mir, BasicBlock, START_BLOCK};
+use mir::cache::Cache;
+
+impl MirCfg {
+    pub fn new<'a, 'tcx>(mir: &Mir, cache: &Cache) -> Self {
+        MirCfg {
+            predecessors: cache.predecessors(mir).clone(),
+            successors: calculate_successors(mir),
```

Comment:
Hmm; I expected to implement the graph trait directly on `Mir`, versus having this intermediate type. Is the idea primarily to have a frozen snapshot for use when computing results, or were there other reasons to introduce `MirCfg`?